Honour suggested backoff time #95 #121

NeilRashbrook opened the pull request:
I couldn't get this to trigger using the STR in #121 but what I was able to do was as follows:
- Delete `~/.mustang/mail.db`
- Set up the account when prompted
- Click on a miscellaneous folder in the account itself
- Click on All Accounts

This seems to trigger a list of every single folder in the account in parallel, thus triggering the server busy error. (This doesn't happen after a restart; you have to do it in the same session.)

benbucksch commented on app/logic/Mail/EWS/EWSAccount.ts:
```diff
@@ -203,7 +203,20 @@ export class EWSAccount extends MailAccount {
+        if (ex.type != "ErrorServerBusy") {
+          throw ex;
+        }
+        let ms = ex.error.MessageXml?.Value?.Value;
```
```suggestion
        let milliseconds = ex.error?.MessageXml?.Value?.Value;
```

benbucksch commented on app/logic/Mail/EWS/EWSAccount.ts:
```diff
@@ -203,7 +203,20 @@ export class EWSAccount extends MailAccount {
+      try {
+        return this.checkResponse(response, aRequest);
+      } catch (ex) {
+        if (ex.type != "ErrorServerBusy") {
```
The Error message says 
`'ConcurrentSyncCalls', policy value '5',`

Can you please:
* Consider the generic `ErrorServerBusy`, and do something sensible, even if we cannot parse the values?
* Please also check for 'ConcurrentSyncCalls' before parsing the values?
* Also parse the value 5. Given the context, I assume them to mean that they accept only 5 parallel calls. That should be something built into `callEWS()`, a throtteling to not have too many  concurrent calls. The default could be something like 10, and then if the server complaints with this error message, we honor this value and do only 5 parallel calls.
* Good idea to also honor the backoff time, as you're doing below. But there should be a sensible upper limit. I've seen backoffs from Office365 that are 1 hour or 1 day. Obviously, if we were to honor that, the app would simply hang for the end user. So, the maximum backup that we honor should be something like 20 seconds.

NeilRashbrook replied:
I don't know what you mean by "parsing the values". The backoff time is explicitly provided as a separate XML element; I'm not parsing it from the message.

benbucksch replied:
> I don't know what you mean by "parsing the values".

1. Backoff time (as you already do)
2. max number of concurrent calls, in the error message: `policy value '5'`, see my last comment point 3. (we don't check that at all: neither read the value, nor have a throttle of this kind. but we need it.)

NeilRashbrook replied:
>     1. Backoff time (as you already do)

No, I don't parse the backoff time.

NeilRashbrook replied:
`error` has a number of child fields:
- ResponseClass (always "Error" by the time we get here)
- MessageText (localised error message)
- ResponseCode (always "ErrorServerBusy" by the time we get here)
- DescriptiveLinkKey (not actually used by Exchange; always 0)
- MessageXml (contains additional information that depends on the error)

In the case of an ErrorServerBusy this last field then contains a Value which contains a Value which is the back off time in milliseconds, but it can contain other information e.g. if you try to write to a read-only field then it will contain the field that you were trying to write.

benbucksch replied:
The full error message is:
`Error: The server cannot service this request right now. Try again later., This operation exceeds the throttling budget for policy part 'ConcurrentSyncCalls', policy value '5',  Budget type: 'Ews'.  Suggested backoff time 5000 ms`

If the message contains `'ConcurrentSyncCalls'`, can we please pick that `policy value '5'` out of there and parse it into an integer and use it as max concurrentSyncCalls?

benbucksch replied:
I understand that this is ugly, but this is Microsoft. It with the microsoft.com domain.

benbucksch commented:
@NeilRashbrook This is low priority. Please stay on #118 for now.

benbucksch commented on app/logic/Mail/EWS/EWSAccount.ts:
```diff
@@ -203,7 +203,24 @@ export class EWSAccount extends MailAccount {
+          milliseconds = 20000; // Sensible upper limit
+        }
+        console.log(`Server busy, using ${milliseconds}ms backoff time`);
+        await new Promise(resolve => setTimeout(resolve, milliseconds));
```
```suggestion
        `sleep(seconds)`;
```

NeilRashbrook replied:
You can't sleep in the UI thread.

benbucksch replied:
I was referring to our `sleep()` function in util.js, which does exactly what you do here.

benbucksch commented on app/logic/Mail/EWS/EWSAccount.ts:
```diff
@@ -203,7 +203,24 @@ export class EWSAccount extends MailAccount {
+      try {
+        return this.checkResponse(response, aRequest);
+      } catch (ex) {
+        if (ex.type != "ErrorServerBusy") {
```
Could you please move this whole throttling into 1-2 functions?

NeilRashbrook commented:
> I was referring to our sleep() function in util.js, which does exactly what you do here.

(Except for being in seconds, meaning that all of the values get round-tripped from and back to milliseconds.)

Anyway, I've implemented maximum concurrency, and added a limit of 50 requests per second (I don't know an easy way to test the latter though).

benbucksch commented on app/logic/Mail/EWS/EWSAccount.ts:
```diff
@@ -198,12 +201,58 @@ export class EWSAccount extends MailAccount {
+      throw ex;
+    }
+    // Check whether the ConcurrentSyncCalls policy was exceeded.
+    if (/'ConcurrentSyncCalls'/.test(ex.message) && /'(\d+)'/.test(ex.message) && Number(RegExp.$1) < this.maxConcurrency) {
```
Can you make the second regexp more specific, please? To avoid that there's another `'4'` before `ConcurrentSyncCalls`. Like:

```suggestion
    if (/'ConcurrentSyncCalls'.*'(\d+)'/.test(ex.message) && Number(RegExp.$1) < this.maxConcurrency) {
```

benbucksch commented on app/logic/Mail/EWS/EWSAccount.ts:
```diff
@@ -198,12 +201,58 @@ export class EWSAccount extends MailAccount {
+    if (/'ConcurrentSyncCalls'/.test(ex.message) && /'(\d+)'/.test(ex.message) && Number(RegExp.$1) < this.maxConcurrency) {
+      this.maxConcurrency = Number(RegExp.$1);
+    }
+    let milliseconds = ex.error.MessageXml?.Value?.Value;
```
What makes this a number?
```suggestion
    let milliseconds = parseInt(ex.error.MessageXml?.Value?.Value);
```

benbucksch commented on app/logic/Mail/EWS/EWSAccount.ts:
```diff
@@ -198,12 +201,58 @@ export class EWSAccount extends MailAccount {
+
   async callEWS(aRequest: JsonRequest): Promise<any> {
-    let response = await appGlobal.remoteApp.postHTTP(this.url, this.request2XML(aRequest), "text", this.createRequestOptions());
+    let milliseconds = Date.now();
+    this.nextConnectionTime.push(milliseconds + 1000);
+    milliseconds = this.nextConnectionTime.shift() - milliseconds;
+    if (milliseconds > 0) {
+      console.log(`Throttling for ${milliseconds}ms because there were 50 connections in the last second`);
+      await sleep(milliseconds / 1000);
+    }
```
Please make this a generic function

benbucksch replied:
We want to re-use it for OWA

benbucksch commented on app/logic/Mail/EWS/EWSAccount.ts:
```diff
@@ -198,12 +201,58 @@ export class EWSAccount extends MailAccount {
+      console.log(`Throttling for ${milliseconds}ms because there were 50 connections in the last second`);
+      await sleep(milliseconds / 1000);
+    }
+    while (this.connections.length >= this.maxConcurrency) {
+      console.log(`Throttling because there are ${this.maxConcurrency} pending connections`);
+      try {
+        await Promise.race(this.connections);
+      } catch (ex) {
+      }
+    }
```
Please make this another generic function

NeilRashbrook commented:
I've adapted the regex we use for owl (it's a different policy name but otherwise the same idea).
I've put the rate limiting in its own function in `util.ts`.
I haven't done anything with the concurrency because that would cause a race condition (the `add` and `remove` need to be in the same async function where the length is checked).

benbucksch commented on app/logic/util/util.ts:
```diff
@@ -76,6 +76,16 @@ export function fileExtensionForMIMEType(mimetype: string) {
+    await sleep(milliseconds / 1000);
+  }
+}
+
```
util.js is only for very generic utility functions. Basically functions that should really be part of the web platform, but are not yet.

benbucksch commented on app/logic/util/util.ts:
```diff
@@ -76,6 +76,16 @@ export function fileExtensionForMIMEType(mimetype: string) {
   return ".ext";
 }
 
+export async function throttleConnectionsPerSecond(nextConnectionTime: number[]) {
```
This parameter **really** needs documentation. Who populates it? (This function, not the caller, just pass it through) How to init it (array length = number of allowed calls per second, and init all values with 0)
